### PR TITLE
Add systemd fine tuning and memprof extension support

### PR DESCRIPTION
**Pull request by mamedin**
Connects to:

https://github.com/artefactual-labs/ansible-atom/issues/109

### CODE REVIEW

File: defaults/main.yml [hunk 1]
```diff
@@ -249,3 +249,28 @@ atom_drmc_mock: "no"
 #   Binder 0.8: "yes" (assumes ES running on same server)
 #   Binder 0.9: "no"
 atom_drmc_es_scripts: "no"
+
+#
+# Systemd AtoM worker parameters
+#
+
+# Check templates/lib/systemd/system/atom-worker.service
```
Comment (sbreker) [inline defaults/main.yml:257]: Nice work parameterizing this!

File: defaults/main.yml [hunk 2]
```diff
@@ -249,3 +249,28 @@ atom_drmc_mock: "no"
...
+
+atom_worker_systemd_start_limit_interval_sec: "24h"
+atom_worker_systemd_start_limit_burst: "3"
```
Comment (sbreker) [inline defaults/main.yml:260]: These will need to be updated for 13575

File: tasks/php.yml [hunk 3]
```diff
@@ -15,3 +15,59 @@
     dest: "/etc/init/php5-fpm.conf"
     line: "reload signal USR2"
   when: "ansible_distribution_version is version('14.04', '==')"
+
```
Comment (sbreker) [inline tasks/php.yml:18]: Nice!

**Review by sbreker — APPROVED** [review]
Thanks @mamedin!  :+1:

File: tasks/php.yml (hunk 3 above)
Comment (mamedin) [reply tasks/php.yml:18]: Thanks Steve, I'll ask now a CR to the support team :)

File: templates/lib/systemd/system/atom-worker.service [hunk 4]
```diff
@@ -20,21 +20,22 @@ WorkingDirectory={{ atom_path }}
...
 {% endif %}
 {% endif %}
 KillSignal=SIGTERM
-Restart=on-failure
-RestartSec=30
-
+Restart={{ atom_worker_systemd_restart }}
+RestartSec={{ atom_worker_systemd_restart_sec }}
+RestartForceExitStatus={{ atom_worker_systemd_restart_force_exit_status }}
+RestartForceExitStatus={{ atom_worker_systemd_success_exit_status }}
```
Comment (hakamine) [inline templates/lib/systemd/system/atom-worker.service:38]: I guess the last one should be `SuccessExitStatus={{ atom_worker_systemd_success_exit_status }}` ?

File: defaults/main.yml [hunk 5]
```diff
@@ -249,3 +249,28 @@ atom_drmc_mock: "no"
...
+atom_worker_systemd_execstart_worker_extra_args: ""
+atom_worker_systemd_restart: "on-failure"
+atom_worker_systemd_restart_sec: "30"
+
+atom_worker_systemd_memory_limit: "-1"
+atom_worker_systemd_error_reporting: "E_ALL"
+
+atom_worker_systemd_restart_force_exit_status: "111"
+atom_worker_systemd_success_exit_status: "111"
+
```
Comment (hakamine) [inline defaults/main.yml:271]: I have a question about the value assigned to `atom_worker_systemd_success_exit_status`. According to the [systemd docs](https://www.freedesktop.org/software/systemd/man/systemd.service.html) :
```
SuccessExitStatus= 
Takes a list of exit status definitions that, when returned by the main service 
process, will be considered successful termination, in addition to the normal 
successful exit status 0 and, except for Type=oneshot, the signals 
SIGHUP, SIGINT, SIGTERM, and SIGPIPE
```
Why is `111` being considered a code for successful termination? In the preceding line, 111 is added to `RestartForceExitStatus=` (i.e., a 111 exit code will cause a service restart), wouldn't it be a contradiction if it is also considered a successful exit code?

**Review by hakamine — CHANGES_REQUESTED** [review]
I noticed 3 things:
- Typo in the systemd atom worker template  (inline comment)
- Question about value in `defaults/main.yml` for `atom_worker_systemd_success_exit_status` (inline comment) (not sure if this needs to be changed; just asking, not sure which is the recommended best practice in this case)
- One of the commit messages has a configuration example, would it be good to move the example to the README instead ? (it will be kind of hard to find it later if only mentioned in the commit message)

File: templates/lib/systemd/system/atom-worker.service (hunk 4 above)
Comment (mamedin) [reply templates/lib/systemd/system/atom-worker.service:38]: Good catch! Thanks :)

File: defaults/main.yml (hunk 5 above)
Comment (mamedin) [reply defaults/main.yml:271]: Hi @hakamine , the main reason to use the `SucessEXitStatus` is to don't show in the worker logs an error when restarting. It is a cosmetic change.

Please, take a look to Steve's Document (Section - Case: atom-worker hits the max-job-count value):

https://docs.google.com/document/d/1sm4gABvoqRouzoUX3CBmRTy8Vg9WwU78XAY00aZ9reY/edit#

When not using `SuccessExitStatus=111`, the logs shows when reaching the `max-job-count` value:

```
Oct 07 03:42:39 vagrant php7.4[4866]: 2021-10-06 20:42:39 > Job worker stopped.
Oct 07 03:42:39 vagrant systemd[1]: atom-worker.service: Main process exited, code=exited, status=111/n/a
Oct 07 03:42:39 vagrant systemd[1]: atom-worker.service: Failed with result 'exit-code'.
Oct 07 03:42:44 vagrant systemd[1]: atom-worker.service: Scheduled restart job, restart counter is at 1.
```

But when using `SuccessExitStatus=111`:

```
Oct 07 03:45:58 vagrant php7.4[5083]: 2021-10-06 20:45:58 > Job worker stopped.
Oct 07 03:45:58 vagrant systemd[1]: atom-worker.service: Succeeded.
Oct 07 03:46:03 vagrant systemd[1]: atom-worker.service: Scheduled restart job, restart counter is at 1.
Oct 07 03:46:03 vagrant systemd[1]: Stopped AtoM worker.
Oct 07 03:46:03 vagrant systemd[1]: Started AtoM worker.
```

**Comment by mamedin** [comment]
@hakamine changes done.

1) Fixed typo on `SuccessExitStatus`
2) README file updated
3) Response about the `atom_worker_systemd_success_exit_status` question here: https://github.com/artefactual-labs/ansible-atom/pull/110#discussion_r854298452

Thanks for suggestions and fixes  :)

**Review by hakamine — APPROVED** [review]
Thank you @mamedin

**Pull request merged**